bpf/lib/overloadable_xdp.h

```diff
@@ -167,12 +177,29 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
 		   __u32 node_id __maybe_unused,
 		   __u32 seclabel __maybe_unused,
 		   __u32 dstid __maybe_unused,
-		   __u32 vni __maybe_unused, int *ifindex __maybe_unused)
+		   __u32 vni __maybe_unused,
+		   void *opt __maybe_unused,
+		   __u32 opt_len __maybe_unused,
+		   bool is_ipv6 __maybe_unused,
+		   int *ifindex __maybe_unused)
 {
 	ctx_store_meta(ctx, CB_ENCAP_NODEID, bpf_ntohl(node_id));
 	ctx_store_meta(ctx, CB_ENCAP_SECLABEL, seclabel);
 	ctx_store_meta(ctx, CB_ENCAP_DSTID, dstid);
 	ctx_set_xfer(ctx, XFER_PKT_ENCAP);
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+	if (opt) {
+		if (!is_ipv6) {
+			__be32 *addr = opt + sizeof(struct geneve_opt_hdr);
+			__be16 *port = opt + sizeof(struct geneve_opt_hdr) + sizeof(__be32);
```

Comment:
Is there a reason this is doing pointer arithmetic with a `void *` rather than using proper structures? :thinking:

Reply:
That's because `ctx_set_encap_info`  receives either geneve_dsr_opt4 or geneve_dsr_opt6. With TC, we can just pass it to `skb_set_tunnel_opt` helper, but we need to retrieve addr and port and store them in the data_meta when it comes to XDP.